### DISCUSSION TITLE
Proposal: New SwitchMode 16 (inverted 15)

### DISCUSSION THREAD

**Discussion by fdcastel** [arendst/Tasmota#17028] [Ideas]
Several `SwitchModes` have their dual for inverted state. E.g.:

```
1 = follow (0 = off, 1 = on)
2 = inverted follow (0 = on, 1 = off)

3 = pushbutton (default 1, 0 = toggle)
4 = inverted pushbutton (default 0, 1 = toggle)

5 = pushbutton with hold (default 1, 0 = toggle, Hold = hold)
6 = inverted pushbutton with hold (default 0, 1 = toggle, hold = hold)

9 = multi change follow (0 = off, 1 = on, 2x change = hold)
10 = inverted multi change follow (0 = on, 1 = off, 2x change = hold)

11 = pushbutton with dimmer mode
12 = inverted pushbutton with dimmer mode

13 = pushon mode (1 = on, switch off using PulseTime)
14 = inverted pushon mode (0 = on, switch off using PulseTime)
```

Except `SwitchMode 15`:

```
15 = send only MQTT message on switch change (Example tele/tasmota/SENSOR = {"Time":"2021-01-01T00:00:00","Switch1":"OFF"})
```

**Proposal:**

```
16 = send only MQTT message on switch change with inverted state (Example tele/tasmota/SENSOR = {"Time":"2021-01-01T00:00:00","Switch1":"ON"})
```

**Rationale:**

 - Makes thing simpler in cases when you need an inverted state. 
 - It's good design to keep it aligned with other SwitchModes.

**Attempted workaround:**

I tried to solve my problem reversing the default `StateText` values with

```
StateText1 ON
StateText2 OFF
```

But this opened a can of worms -- the return values for `SetOptionXXX` commands are also affected and broke my management scripts.

**Comment by lordbonne** [comment]
This is exactly what I need right now. +1 :)

**Comment by arendst** [comment]
I didn't write this part of the code and haven't looked at it yet but if I act on the text

> send only MQTT message on switch change

I suppose it doesn't matter if the switch is turned on or off. In either case I would expect to see a MQTT message. What needs to be inverted here?

**Reply by arendst** [reply]
Ah, I did a test and I see what you mean.

Will add the inverted option.

**Comment by arendst** [comment]
Implemented in latest dev release

**Comment by stacklover** [comment]
I have 12.2.0.6(sensors) running on ESP8285N08 and see no functional difference between switchmode 15 and 16. In both cases GPIO at GND level will send "OFF". Here only switch 1 is closed and thus GPIO is at GND potential:
generator/tele/SENSOR = {"Time":"2022-12-06T11:59:28","Switch1":"OFF","Switch2":"ON","Switch3":"ON","Switch4":"ON"}

**Reply by sfromis** [reply]
On Discord, you also posted:
> weird.. after I set switchmode1 15 to verify it works the same, then setting back to switchmode1 16 it works correctly

**Reply by stacklover** [reply]
that was an observational error. I did not account for the delay in seeing the message in the console which comes a few seconds after I change the switch. Changing fast enough it seemed to work. But after I tried all again with more relaxed waiting I verified that there is no difference between 15 and 16. It should be easy to reproduce by you.

**Reply by stacklover** [reply]
Without changing the switch it is easy to see here:

14:58:28.404 MQT: generator/tele/STATE = {"Time":"2022-12-06T14:58:28","Uptime":"0T03:23:04","UptimeSec":12184,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"LIGHTBULB","BSSId":"D8:0D:17:B7:5D:B1","Channel":11,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:03"}}
14:58:28.412 MQT: generator/tele/SENSOR = {"Time":"2022-12-06T14:58:28","Switch1":"ON","Switch2":"ON","Switch3":"ON","Switch4":"ON"}
14:58:36.917 CMD: switchmode1 15
14:58:36.923 MQT: generator/stat/RESULT = {"SwitchMode1":15}
14:58:38.412 MQT: generator/tele/STATE = {"Time":"2022-12-06T14:58:38","Uptime":"0T03:23:14","UptimeSec":12194,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"LIGHTBULB","BSSId":"D8:0D:17:B7:5D:B1","Channel":11,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:03"}}
14:58:38.421 MQT: generator/tele/SENSOR = {"Time":"2022-12-06T14:58:38","Switch1":"ON","Switch2":"ON","Switch3":"ON","Switch4":"ON"}
14:58:48.439 MQT: generator/tele/STATE = {"Time":"2022-12-06T14:58:48","Uptime":"0T03:23:24","UptimeSec":12204,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"LIGHTBULB","BSSId":"D8:0D:17:B7:5D:B1","Channel":11,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:03"}}
14:58:48.448 MQT: generator/tele/SENSOR = {"Time":"2022-12-06T14:58:48","Switch1":"ON","Switch2":"ON","Switch3":"ON","Switch4":"ON"}
14:58:49.546 CMD: switchmode1 16
14:58:49.553 MQT: generator/stat/RESULT = {"SwitchMode1":16}
14:58:58.452 MQT: generator/tele/STATE = {"Time":"2022-12-06T14:58:58","Uptime":"0T03:23:34","UptimeSec":12214,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"LIGHTBULB","BSSId":"D8:0D:17:B7:5D:B1","Channel":11,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:03"}}
14:58:58.461 MQT: generator/tele/SENSOR = {"Time":"2022-12-06T14:58:58","Switch1":"ON","Switch2":"ON","Switch3":"ON","Switch4":"ON"}

Just to be sure, I also tried with restarting the system (power off/on)...

**Comment by host-felice** [comment]
I upgraded to 12.3.0 and switchmode 16 doesn't work as inverted :-( it does the same as switchmode 15. I did a restart just to be sure, but same result. I did the upgrade and tested it with 2 devices.

**Comment by host-felice** [comment]
I tried 12.3.1 and still the same problem: switchmode 15 and 16 do work the same way, switchmode 16 is not inverted.

**Comment by JustinWrong1** [comment]
Just updated device to 12.3.1, still find no difference between SwitchMode 15 and 16. It does not invert the status. Does anyone know if there is progress on this issue? TIA

**Reply by sfromis** [reply]
Doing a trivially simple test, I saw a difference. Both Switch1 and Switch2 are standard buttons, connecting to ground when pushed, otherwise pulled up.
```
SwitchMode1 15
SwitchMode2 16
SetOption114 off

{"Time":"2022-12-28T23:50:54.892+01:00","Switch1":"OFF","Switch2":"ON"}
{"Time":"2022-12-28T23:50:58.352+01:00","Switch1":"ON","Switch2":"ON"}

{"Time":"2022-12-28T23:51:06.525+01:00","Switch1":"ON","Switch2":"ON"}
{"Time":"2022-12-28T23:51:09.435+01:00","Switch1":"ON","Switch2":"OFF"}
```
What I did was first pushing and releasing Switch1, then the same for Switch2. Clearly the resulting messages are different, with inverse states.

**Reply by stacklover** [reply]
true. But in the next "teleperiod"ically sent SENSOR message the switches are not inverted again.
When I look at the strange solution for this invering in the source, I can see why this is: by sending the message before updating the value, this works once when the switch really changes, but not in subsequent SENSOR messages...

**Reply by stacklover** [reply]
In https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_support/support_switch.ino

          }
          break;
        case PUSH_IGNORE:
          Switch.last_state[i] = button;                        // Update switch state before publishing
          MqttPublishSensor();
          break;
        case PUSH_IGNORE_INV:
          MqttPublishSensor();                                  // Publishing before update
          break;
        }
        Switch.last_state[i] = button;
      }
      if (switchflag <= POWER_TOGGLE) {

**Reply by stacklover** [reply]
Sorry that I did not clarify this earlier - only after entering the simple test as above I saw the problem now and dug deep into the source code...

**Reply by arendst** [reply]
Try latest dev. (https://github.com/arendst/Tasmota/commit/ff3d90c36d13cc412d388e26dfa1b4aa0ebdb19b)

**Reply by stacklover** [reply]
Well... Now the instantly sent Message is wrong and the "Teleperiod" Message is correct...
This is also clear from the new source - now it is double inverted when it is instantly sent.

Would it not be more wise to have a "Switch_i" and "Switch_in" in the configuration - equivalent to how Butons, Relays and LEDs are handled? Naturally I don't know how much effort that is to do and whether that might break other stuff...

**Reply by stacklover** [reply]
}
          break;
        case PUSH_IGNORE:
        case PUSH_IGNORE_INV:
          Switch.last_state[i] = button;                        // Update switch state before publishing
          MqttPublishSensor();
          break;
        }
        Switch.last_state[i] = button;
      }
      if (switchflag <= POWER_TOGGLE) {

**Reply by stacklover** [reply]
above fixes the current solution...
I tested it.

**Reply by JustinWrong1** [reply]
> Would it not be more wise to have a "Switch_i" and "Switch_in" in the configuration - equivalent to how Butons, Relays and LEDs are handled? Naturally I don't know how much effort that is to do and whether that might break other stuff...

This would be very handy for people who understand the basics.

**Reply by JustinWrong1** [reply]
> Try latest dev. ([ff3d90c](https://github.com/arendst/Tasmota/commit/ff3d90c36d13cc412d388e26dfa1b4aa0ebdb19b))

Thanks for your efforts Mr Arends, being a complete novice, and managing the standard Tosmotizer methods of flashing devices, Is there perhaps a tutorial to help me action the above implantation version? your assistance is greatly appreciated.

**Reply by sfromis** [reply]
You can get the precompiled binary for the new development build doing an upgrade via the URL from http://ota.tasmota.com/tasmota/ (no download first). Probably `http://ota.tasmota.com/tasmota/tasmota.bin.gz`

**Reply by stacklover** [reply]
Warning: above mentioned fix is not implemented yet. The current (as of now) development tasmota will still exhibit the problem!

**Reply by arendst** [reply]
It is now https://github.com/arendst/Tasmota/commit/32a51da43d064447a5e1d0b1c67990de7f8945a9

**Comment by host-felice** [comment]
Switchmode 16. I did close the pin to GND (pin was in pull-up): rsl message is wrong (should be ON since inverted), teleperiod is ok. I didn't open the pin-gnd while typing teleperiod.

16:18:05.010 RSL: SENSOR = {"Time":"2022-12-30T16:18:05","Switch1":"OFF"}
16:18:12.142 CMD: teleperiod
16:18:12.146 RSL: RESULT = {"TelePeriod":300}
16:18:12.395 RSL: STATE = [CUT BECAUSE NOT INTERESTING]
16:18:12.401 RSL: SENSOR = {"Time":"2022-12-30T16:18:12","Switch1":"ON"}
